futures-executor/src/enter.rs

```diff
@@ -0,0 +1,96 @@
+use std::prelude::v1::*;
+use std::cell::{RefCell, Cell};
+use std::fmt;
+
+thread_local!(static ENTERED: Cell<bool> = Cell::new(false));
+
+/// Represents an executor context.
+///
+/// For more details, see [`enter` documentation](fn.enter.html)
+pub struct Enter {
+    on_exit: RefCell<Vec<Box<Callback>>>,
+    permanent: bool,
+}
+
+/// An error returned by `enter` if an execution scope has already been
+/// entered.
+#[derive(Debug)]
+pub struct EnterError {
+	_a: (),
+}
+
+/// Marks the current thread as being within the dynamic extent of an
+/// executor.
+///
+/// Executor implementations should call this function before blocking the
+/// thread. If `None` is returned, the executor should fail by panicking or
+/// taking some other action without blocking the current thread. This prevents
+/// deadlocks due to multiple executors competing for the same thread.
+///
+/// # Error
+///
+/// Returns an error if the current thread is already marked
+pub fn enter() -> Result<Enter, EnterError> {
+    ENTERED.with(|c| {
+        if c.get() {
+            Err(EnterError { _a: () })
+        } else {
+            c.set(true);
+
+            Ok(Enter {
+                on_exit: RefCell::new(Vec::new()),
+                permanent: false,
+            })
+        }
+    })
+}
+
+impl Enter {
+    /// Register a callback to be invoked if and when the thread
+    /// ceased to act as an executor.
+    pub fn on_exit<F>(&self, f: F) where F: FnOnce() + 'static {
```

Comment:
This seems really fragile to me-- do you have a use-case in mind for it?

Reply:
I believe this was intended for things along the lines of `handle.set_default_for(&enter)` where you'd have a handle to some "thing" (like an event loop, timer, etc) and you, as the executor, would want to configure these global resources for your particular scope. For example you'd say something like "use this timer when you're on this executor" or "use that event loop". 

The `on_exit` here was a way for libraries to use `Enter` to communicate this sort of "per executor" state where you'd have a function that takes an `Enter` which sets up some state for your thread and is then cleaned up in `Enter::on_exit`. An example of this would be you'd register an event loop handle in a thread local, and then when the executor finishes you'd pull out that event loop handle and release it.